clang/include/clang/CIR/Dialect/IR/CIROps.td

```diff
@@ -1415,6 +1415,113 @@ def VTableAddrPointOp : CIR_Op<"vtable.address_point",
   let hasVerifier = 1;
 }
 
+//===----------------------------------------------------------------------===//
+// SetBitfieldOp
+//===----------------------------------------------------------------------===//
+
+def SetBitfieldOp : CIR_Op<"set_bitfield"> {
+  let summary = "Set a bitfield";
+  let description = [{ 
+    The `cir.set_bitfield` operation provides a store-like access to 
+    a bit field of a record.
+
+    It expects an address of a storage where to store, a value being stored,
+    a type of a bitfield, a name of a bit field, a pointer to a storage in the
+    base record, a size of the storage, a size the bit field, an offset 
+    of the bit field and a sign. Returns a value being stored.
+
+    Example:
+    ```mlir
+    // Suppose we have a struct with multiple bitfields stored in
+    // different storages. The `cir.set_bitfield` operation sets the value
+    // of the bitfield
+
+    // Sets the bitfield with name a1
+    %3 = cir.set_bitfield %1, %2(<name = "a1", elt_type = !u16i, storage_size = 16, size = 15, offset = 0, is_signed = true>) : (!s32i, !cir.ptr<!u16i>) -> !s32i
```

Comment:
Storage size doesn't really need be maintained, we can derive from the `elt_type`. The attribute could provide an extra method class with `getStorageSize()`, that is computed directly on top of datalayout.

For instance, if you see how it's created: `Info.StorageSize = getSizeInBits(StorageType).getQuantity();`, where `getSizeInBits` gets implemented in terms of `return charUnits::fromQuantity(dataLayout.layout.getTypeSizeInBits(Ty));`. Such extra method can just query datalayout the same).

Reply:
well, I fixed this several times and rolled back all the changes again) because there is kind of mess connected with `volatile` support here. But I think it's ok for now - we infer the storage type according to the `volatile` qualifier and looks like it's enough, so we wan't do it for the storage size.